common/check_internal.h

```diff
@@ -5,73 +5,128 @@
-};
+}
+#endif
 
 }  // namespace Carbon::Internal
 
-// Raw exiting stream. This should be used when building forms of exiting
-// macros. It evaluates to a temporary `ExitingStream` object that can be
-// manipulated, streamed into, and then will exit the program.
-#define CARBON_CHECK_INTERNAL_STREAM() \
-  Carbon::Internal::ExitingStream::Helper() | Carbon::Internal::ExitingStream()
+// Dispatch helper between debug and optimized builds for the different
+// formulation of the actual check function.
+#ifndef NDEBUG
+// Pass the format string as a runtime parameter in debug builds.
+#define CARBON_INTERNAL_CHECK_IMPL_CALL(kind, file, line, condition_str, \
+                                        format_str, ...)                 \
+  Carbon::Internal::CheckFail(kind, file, line, condition_str,           \
+                              format_str __VA_OPT__(, ) __VA_ARGS__)
+#else
+// Pass the format string as a template parameter in optimized builds. Note that
+// the template arguments include commas and so require extra parentheses to use
+// within a macro.
+#define CARBON_INTERNAL_CHECK_IMPL_CALL(kind, file, line, condition_str,     \
+                                        format_str, ...)                     \
+  (Carbon::Internal::CheckFail<kind, file, line, condition_str, format_str>( \
+      __VA_ARGS__))
+#endif
+
+// Calls `CheckFail` with the provided message plus a newline and with no
+// parameters. This implements check messages without any formatted values.
+#define CARBON_INTERNAL_CHECK_IMPL(kind, file, line, condition_str) \
+  CARBON_INTERNAL_CHECK_IMPL_CALL(kind, file, line, condition_str, "")
+
+// Calls `CheckFail` with the provided message prefix concatenated with a format
+// string and a newline. All other parameter are passed as arguments to
+// `CheckFail` to be rendered into the format string.
+#define CARBON_INTERNAL_CHECK_IMPL_FORMAT(kind, file, line, condition_str, \
+                                          format_str, ...)                 \
+  CARBON_INTERNAL_CHECK_IMPL_CALL(kind, file, line, condition_str,         \
+                                  format_str __VA_OPT__(, ) __VA_ARGS__)
+
+// Implements the failure of a check.
+//
+// Forms a message prefix for the failure that includes its source location and
+// the failing condition, then adds any format string and formatted arguments.
+#define CARBON_INTERNAL_CHECK(condition, ...)      \
+  CARBON_INTERNAL_CHECK_IMPL##__VA_OPT__(_FORMAT)( \
+      "CHECK", __FILE__, __LINE__, #condition __VA_OPT__(, ) __VA_ARGS__)
+
+// Implements the fatal macro.
+//
+// Forms a message prefix for the failure that includes its source location,
+// then adds any format string and formatted arguments.
+#define CARBON_INTERNAL_FATAL(...)                 \
+  CARBON_INTERNAL_CHECK_IMPL##__VA_OPT__(_FORMAT)( \
+      "FATAL", __FILE__, __LINE__, "" __VA_OPT__(, ) __VA_ARGS__)
+
+#ifdef NDEBUG
+// For optimized builds we have a dead check that we want to potentially "use"
```

Comment:
Clarify that this is dead because it is for DCHECK, which is only checking in debug builds.

Reply:
Done.